### PR TITLE
MM-1699 Refactored post handling/updating on both the client and server

### PR DESCRIPTION
**Pull request by jwilander**
Major changes in this PR:
* Refactored post_list.jsx to have a much better and cleaner implementation
* Added a new API service for getting posts updated/created since a certain time
* Refactored RHS to have better internal naming and to be in separate files
* Posts now only re-render when the post object has actually changed.
* Post Store on the client merges new post_lists from the server with the old and does not clobber the old stored post_lists

### CODE REVIEW

File: api/post.go [hunk 1]
```diff
@@ -603,6 +602,40 @@ func getPosts(c *Context, w http.ResponseWriter, r *http.Request) {
...
+
+	time, err := strconv.ParseInt(params["time"], 10, 64)
+	if err != nil {
+		c.SetInvalidParam("getPostsSince", "time")
+		return
+	}
+
+	cchan := Srv.Store.Channel().CheckPermissionsTo(c.Session.TeamId, id, c.Session.UserId)
+
+	if !c.HasPermissionsToChannel(cchan, "getPostsSince") {
```
Comment (coreyhulen) [inline api/post.go]: move after Srv.Store.Post().GetPostsSince (otherwise it's serial calls)

File: api/post.go [hunk 2]
```diff
@@ -603,6 +602,40 @@ func getPosts(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	pchan := Srv.Store.Post().GetPostsSince(id, time)
```
Comment (coreyhulen) [inline api/post.go]: do we want to set an upper limit in the query?  1000 items?  or something.  What happens when you've been gone for weeks and go into a busy channel.

Comment (jwilander) [inline api/post.go]: The DB query itself has a limit of 100 (+ relevant parent posts) right now, making it passable as an argument can be done as well

File: api/post.go (hunk 1 above)
Comment (jwilander) [inline api/post.go]: Yup, good catch

File: store/sql_post_store.go [hunk 3]
```diff
@@ -299,6 +299,69 @@ func (s SqlPostStore) GetPosts(channelId string, offset int, limit int) StoreCha
...
+			result.Err = model.NewAppError("SqlPostStore.GetPostsSince", "We couldn't get the posts for the channel", "channelId="+channelId+err.Error())
+		} else {
+
+			list := &model.PostList{Order: make([]string, 0, len(posts))}
+
+			for _, p := range posts {
+				if p.ImgCount > 0 {
+					p.Filenames = []string{}
+					for i := 0; int64(i) < p.ImgCount; i++ {
+						fileUrl := "/api/v1/files/get_image/" + p.ChannelId + "/" + p.Id + "/" + strconv.Itoa(i+1) + ".png"
```
Comment (coreyhulen) [inline store/sql_post_store.go]: it's get_image?  I thought it was just /get/

Comment (jwilander) [inline store/sql_post_store.go]: This is to support legacy systems that used the old images, do we want to get rid of it? I can get rid of it from the other GetPosts DB functions too. Although I think that would be better in a different ticket (I can still remove it here if we want)

Comment (crspeller) [inline store/sql_post_store.go]: I would vote for getting rid of it now while refactoring.

File: web/react/components/create_comment.jsx [hunk 4]
```diff
@@ -57,11 +55,10 @@ module.exports = React.createClass({
 
         PostStore.storePendingPost(post);
         PostStore.storeCommentDraft(this.props.rootId, null);
-        this.setState({messageText: '', submitting: false, postError: null, previews: [], serverError: null});
 
         client.createPost(post, ChannelStore.getCurrent(),
             function(data) {
-                AsyncClient.getPosts(true, this.props.channelId);
+                //AsyncClient.getPosts(true, this.props.channelId);
```
Comment (crspeller) [inline web/react/components/create_comment.jsx]: Why is this only commented out?

Comment (jwilander) [inline web/react/components/create_comment.jsx]: My mistake, fixing..

**Comment by jwilander** [comment]
@crspeller and @coreyhulen ready for re-review.

File: web/react/stores/post_store.jsx [hunk 5]
```diff
@@ -99,8 +99,57 @@ var PostStore = assign({}, EventEmitter.prototype, {
...
+    storePosts: function storePosts(channelId, newPostList) {
+        if (isPostListNull(newPostList)) {
+            return;
+        }
+
+        var postList = PostStore.getPosts(channelId);
+        if (isPostListNull(postList)) {
+            postList = newPostList;
+        } else {
+            postList = makePostListNonNull(postList);
```
Comment (crspeller) [inline web/react/stores/post_store.jsx]: Little bit of paranoia here? We check if it is not null then really make sure of it!

File: web/react/components/rhs_comment.jsx [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            loading = (
+                <img
+                    className='post-loading-gif pull-right'
+                    src='/static/images/load.gif'
+                />
+            );
+        }
+
+        var ownerOptions;
+        if (isOwner && post.state !== Constants.POST_FAILED && post.state !== Constants.POST_LOADING) {
```
Comment (crspeller) [inline web/react/components/rhs_comment.jsx]: Lots of this is duplicate code from post_info.jsx. Probably should be refactored so we don't have duplicate code.

File: web/react/stores/post_store.jsx (hunk 5 above)
Comment (jwilander) [inline web/react/stores/post_store.jsx]: Haha yeah, the real reason I did this was to skip the merge code (which was at first more complicated) if it's a totally new post list. I'm going to change it to be more reasonable

File: web/react/components/rhs_comment.jsx (hunk 6 above)
Comment (jwilander) [inline web/react/components/rhs_comment.jsx]: I agree although I think it should be done in a separate ticket, I don't want this PR to be too much of a monster just before release.

Comment (crspeller) [inline web/react/components/rhs_comment.jsx]: Sounds good. When you create the ticket post it here.

**Comment by crspeller** [comment]
Travis is OK here. Only the push one failed which is the unmerged code without the build fix.

**Comment by crspeller** [comment]
Code looks good.

**Comment by rgarmsen2295** [comment]
+1 from me as well, really looking forward to the performance improvements

File: web/react/components/rhs_comment.jsx (hunk 6 above)
Comment (jwilander) [inline web/react/components/rhs_comment.jsx]: It's MM-2038 on Jira.

**Comment by crspeller** [comment]
Waiting for @coreyhulen to re-review

**Comment by crspeller** [comment]
@jwilander If perfect scroll-bar is gone you should remove it from head.html (any anywhere else it might be lurking)

**Comment by crspeller** [comment]
Merged so we can get to testing before release.

**Pull request merged**